### PR TITLE
Fix Chatbox scroll to latest and add fetch more (pagination)

### PR DESCRIPTION
**Pull request by ahuang11**
UPDATED comments:
Pretty happy that I found some solutions!

https://github.com/holoviz/panel/assets/15331990/4b132fad-63fe-48b1-8072-b56cdb19001c

To fix scroll, I explicitly set height (or use the provided height/max_height). This significantly reduces complexity since there's only one case to cover instead of getting the window / template main / or column. 

To fix the timing issue mentioned [here](https://github.com/holoviz/panel/issues/5137#issuecomment-1612221063), I used Python callback to call Javascript.

Doing this also makes the chatbox look and feel better in my opinion (the text input isn't jumping around).

To fix performance, I added pagination.

---
OLD comments:
Apparently, this is way more difficult than expected with many edge cases.

1. with/without template
2. fixed height
3. nesting within column

I would like some help on this to work on this together. The current issues:
1. scroll happens immediately, before it gets appended so it only scrolls to the second latest message, rather than the latest
2. if ascending, the text box bounces back to the top
3. how to handle when chatbox isn't the only thing in the template

```
import panel as pn

pn.extension(sizing_mode="stretch_width", layout_compatibility='error')

value = [{"User": c} for c in "ABCDEFGHIJKLMOOPQRSTUVWXYZ"]
chat_box = pn.widgets.ChatBox(
    value=value,
    primary_name="User",
    height=500,
)
(chat_box).servable()
```

### CODE REVIEW

File: panel/widgets/chatbox.py [hunk 1]
```diff
@@ -266,58 +272,102 @@ def __init__(self, **params):
...
         self._composite[:] = box_objects
 
         # add interactivity
         self.param.watch(self._refresh_log, "value")
+        self._fetch_button.on_click(self._fetch_more)
+        self._scroll_button.on_click(self._trigger_scroll_callback)
+        self._hidden_scroll_trigger.jscallback(
+            value="""
+                const outerColumn = document.querySelector(".bk-Column");
+                const column = outerColumn.shadowRoot.querySelector(".bk-Column");
```
[the remark this answers is not in the document]
Comment (ahuang11) [reply panel/widgets/chatbox.py:361]: Oh yeah, I already implemented `css_classes=layout.get("css_classes", []) + ["chat-log"]` and was planning to use that to narrow down the result.

However, I'm having issues just doing `document.querySelector(".bk-Column.chat-log")` which returns null. I thought that would recursively find it, but I suspect the shadowRoot is hiding it?

[the remark this answers is not in the document]
Comment (ahuang11) [reply panel/widgets/chatbox.py:361]: I also tried to inject chat_log as an arg:
```
        self._hidden_scroll_trigger.jscallback(
            value="""
                console.log(chat_log)
            """,
            args={"ascending": self.ascending, "chat_log": self._chat_log},
        )
```
But it seems to be attributes and not the html element; not sure how to back out the xref ID
image

[the remark this answers is not in the document]
Comment (philippjfr) [reply panel/widgets/chatbox.py:361]: I don't think implementing this using jscallbacks for this is ever going to be a reasonable approach. I think if we really want this it'll have to be using a custom Bokeh model or ReactiveHTML component.

**Comment by ahuang11** [comment]
Closing in favor of https://github.com/holoviz/panel/pull/5333

**Pull request closed without merging**